presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RuntimeReorderJoinSides.java

```diff
@@ -66,9 +76,35 @@ public Result apply(JoinNode joinNode, Captures captures, Context context)
             return Result.empty();
         }
 
-        StatsProvider statsProvider = context.getStatsProvider();
-        double leftOutputSizeInBytes = statsProvider.getStats(joinNode.getLeft()).getOutputSizeInBytes(joinNode.getLeft().getOutputVariables());
-        double rightOutputSizeInBytes = statsProvider.getStats(joinNode.getRight()).getOutputSizeInBytes(joinNode.getRight().getOutputVariables());
+        double leftOutputSizeInBytes = Double.NaN;
+        double rightOutputSizeInBytes = Double.NaN;
+        PlanNode actualLeft = context.getLookup().resolve(joinNode.getLeft());
+        PlanNode actualRight = context.getLookup().resolve(joinNode.getRight());
+        if (actualLeft instanceof TableScanNode && actualRight instanceof ExchangeNode && actualRight.getSources().size() == 1
+                && context.getLookup().resolve(actualRight.getSources().get(0)) instanceof TableScanNode) {
+            // Simple plan is characterized as Join[leftSrc: TableScan, rightSrc: Exchange[src: TableScan]].
+            // For simple plans, directly fetch the overall table sizes as the size of the join sides to have
+            // accurate input bytes statistics and meanwhile avoid non-negligible cost of collecting and processing
+            // per-column statistics.
+            TableScanNode leftScan = (TableScanNode) actualLeft;
+            TableScanNode rightScan = (TableScanNode) context.getLookup().resolve(actualRight.getSources().get(0));
+            leftOutputSizeInBytes = metadata.getTableStatistics(context.getSession(),
+                    leftScan.getTable(),
+                    ImmutableList.copyOf(leftScan.getAssignments().values()),
+                    new Constraint<>(leftScan.getCurrentConstraint())).getTotalSize().getValue();
+            rightOutputSizeInBytes = metadata.getTableStatistics(context.getSession(),
+                    rightScan.getTable(),
+                    ImmutableList.copyOf(rightScan.getAssignments().values()),
+                    new Constraint<>(rightScan.getCurrentConstraint())).getTotalSize().getValue();
+        }
+
+        // Fall back using statsProvider to estimate left and right output size.
+        if (Double.isNaN(leftOutputSizeInBytes) || Double.isNaN(rightOutputSizeInBytes)) {
```

Comment:
what if instead of having these two separate paths, we just use the regular stats calculator path, but added an Optional<Double> totalSize (or NaN for absent cases) to PlanNodeStatsEstimate.  then if we call getOutputsizeInBytes()with all of the input variables, then we return that number, and otherwise we calculate as usual per-column. We'd have to handle it correctly in the stats calculators to populate it correctly.  But it would also allow non-materialized queries to benefit from this change.

@arhimondr what do you think?

Reply:
Yeah, adding totalSize field also in `PlanNodeStatsEstimate` will merge the current "fast-path" into the regular statsCalculator path.

My only concern is whether it worth adding the field. The downside is only a small portion will have actual values because operations like Aggregate, Join, Project, ... will easily make it NaN. Meanwhile, the advantage seems quite narrow, because for non-materialized queries it will only be effective for directly joining two actual tables. This seems a bit too-good-to-be-true because the optimizer tends to pushdown whatever filter/projection/otherOps possible below the join right?

Reply:
It also feels to me that it should be the responsibility of the `getOutputSizeInBytes` method to return the size based on the column size information if the `totalSize` is not set. That will effectively merge these two paths and make the code a bit cleaner. 

I'm a little confused why do we think this approach is not preferred? Could you please elaborate a little bit more on that?

Reply:
@arhimondr it's not possible (or at least easy) to compute stats based on columns if you don't have a list of columns because you won't know how many columns don't have stats that you need to account for.  In order to handle this, we'd need to change the StatsNormalizer not to remove unknown stats from the list of variable statistics. This is what i was asking about before.